setup.go

```diff
@@ -165,11 +173,45 @@ func setup(c *caddy.Controller) error {
 		}
 	}
 
-	fp.httpTransport.DialContext = (&net.Dialer{
-		Timeout:   fp.dialTimeout,
-		KeepAlive: 30 * time.Second,
-		DualStack: true,
-	}).DialContext
+	if fp.upstream != "" {
+		fixedURL, err := url.Parse(fp.upstream)
+		if err != nil {
+			return c.ArgErr()
+		}
+		switch fixedURL.Scheme {
+		case "http":
+			fp.httpTransport.Proxy = http.ProxyURL(fixedURL)
+			fallthrough
+		default:
+			dialer := &net.Dialer{
+				Timeout:   fp.dialTimeout,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}
+			newDialer, err := proxy.FromURL(fixedURL, dialer, nil)
+			if err != nil {
+				return c.ArgErr()
+			}
+			fp.dial = newDialer.Dial
+			fp.httpTransport.Dial = newDialer.Dial
+		}
+	}
+
+	if fp.httpTransport.Dial == nil {
```

Comment:
Could be in `else` block, corresponding to `if fp.upstream != ""`, making code easier to follow